### PR TITLE
Interlinks Autocomplete: Mark III

### PR DESCRIPTION
**Pull request by jleandroperez**
### Details
In this PR we're wiring the Interlinks Autocomplete UI, so that whenever the user selects a row, an Interlink for the selected note gets inserted.

cc @eshurakov Yet another piece of the puzzle!
Thanks in advance!!

Ref. #914

### Test
1. Open any note
2. Type `[keyword` with `keyword` being a word contained within any of your Note's title

- [x] Verify that selecting any of the rows that show up causes the App to insert an Interlink to the target note
- [x] Verify that **Shake > Undo** reverts the Interlink Insertion OP

### Release
These changes do not require release notes.

### CODE REVIEW

File: Simplenote/Classes/UITextView+Simplenote.swift [hunk 1]
```diff
@@ -28,6 +28,76 @@ extension UITextView {
...
+// MARK: - Updating!
+//
+extension UITextView {
+
+    /// Inserts the specified Text at a given range.
+    /// - Note: Resulting selected range will end up at the right hand side of the newly inserted text.
+    ///
+    func insertText(text: String, in range: Range<String.Index>) {
+        registerUndoCheckpointAndPerform { storage in
+            let range = self.text.utf16NSRange(from: range)
```
Comment (eshurakov) [inline Simplenote/Classes/UITextView+Simplenote.swift]: `NSRange(range, in: self.text)` and can maybe completely remove `utf16NSRange`?

File: Simplenote/Classes/UITextView+Simplenote.swift [hunk 2]
```diff
@@ -28,6 +28,76 @@ extension UITextView {
...
+            storage.replaceCharacters(in: range, with: text)
+            self.selectedRange = NSRange(location: range.location + text.count, length: .zero)
```
Comment (eshurakov) [inline Simplenote/Classes/UITextView+Simplenote.swift]: I believe it is incorrect to add `text.count` to NSRange. `text.count` is based on the representation of the string (utf8 / utf16) and can result in a wrong adjustment.

I would try to do:
```
let insertedTextRange = NSRange(text.fullRange, in: text)
self.selectedRange = NSRange(location: range.location + insertedTextRange.length, length: .zero)
```

File: Simplenote/Classes/SPNoteEditorViewController+Extensions.swift [hunk 3]
```diff
@@ -599,17 +599,22 @@ extension SPNoteEditorViewController: InterlinkProcessorPresentationContextProvi
...
 
-    var interlinkExcudedEntityID: NSManagedObjectID? {
+    func excludedEntityIdentifierForInterlinkProcessor(_ processor: InterlinkProcessor) -> NSManagedObjectID? {
         currentNote?.objectID
     }
+
+    func interlinkProcessor(_ processor: InterlinkProcessor, insert text: String, in range: Range<String.Index>) {
+        noteEditorTextView.insertText(text: text, in: range)
+        processor.dismissInterlinkLookup()
+    }
```
Comment (eshurakov) [inline Simplenote/Classes/SPNoteEditorViewController+Extensions.swift:613]: Idea: do not to pass markdown range to the Processor and back here. Instead we can figure it out right here based on the current selection in the editor.

File: Simplenote/Classes/UITextView+Simplenote.swift (hunk 2 above)
Comment (jleandroperez) [reply Simplenote/Classes/UITextView+Simplenote.swift]: You're 💯 on spot. I need to fix mac as well (this code was borrowed from there!!)

File: Simplenote/Classes/SPNoteEditorViewController+Extensions.swift (hunk 3 above)
Comment (jleandroperez) [reply Simplenote/Classes/SPNoteEditorViewController+Extensions.swift:613]: (For the record!!) Discussed over Slack. Let's keep the Range (for now at least!).

Thank you for bringing this up!!

**Comment by jleandroperez** [comment]
@eshurakov Ready for another peek!! thanks in advance!!

**Review by eshurakov — APPROVED** [review]
Looks great!

:shipit:

**Comment by jleandroperez** [comment]
Thank you sir!!

**Pull request merged**